enhancements/dns/forwarding.md

```diff
@@ -0,0 +1,246 @@
+upstream resolvers to subdomain associations:
+
+```go
+// DNSSpec is the specification of the desired behavior of the DNS.
+type DNSSpec struct {
+	// +optional
+	Upstreams []Upstream `json:"upstreams,omitempty"`
+}
+```
+
+### User Stories
+
+#### Story 1
+
+As a customer with OpenShift running in AWS and connected to my data center by VPN, I want OpenShift DNS to resolve
+name queries for our other internal devices using the DNS servers in our data center.
+
+### Implementation Details/Notes/Constraints
+
+The maximum number of `Nameservers` is 15.
+
+The cluster-dns-operator will generate the `forward` portion of `configmap/dns-default` based on `Upstream` of
+`dnses.operator.openshift.io` instead of `forward` being statically defined. To achieve this:
+
+1. The [default](https://github.com/openshift/cluster-dns-operator/blob/master/assets/dns/configmap.yaml) ConfigMap
+asset and associated code from the [manifests pkg](https://github.com/openshift/cluster-dns-operator/blob/master//pkg/manifests/manifests.go#L70:6)
+should be removed.
+
+2. The [desiredDNSConfigMap](https://github.com/openshift/cluster-dns-operator/blob/master/pkg/operator/controller/controller_dns_configmap.go)
+function must be modified to create a ConfigMap type with a `forward` configuration based on `Upstream`.
+If `Upstream` is not present or an invalid `Upstream` is provided, the ConfigMap will contain
+`forward . /etc/resolv.conf`. Otherwise, desiredDNSConfigMap will use the provided `Upstream` to construct the
+forwarding configuration.
+
+#### Updating
+Changes to `Upstream` will trigger a [rolling update](https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/#performing-a-rolling-update)
+of the CoreDNS DaemonSet.
+
+#### Validation
+`Domain` must conform to the [RFC 1123](https://tools.ietf.org/html/rfc1123#page-13) definition of a subdomain. Each
+nameserver in `Nameservers` must be a valid IPv4 or IPv6 address. If the nameserver listens on a UDP port other than 53,
+a valid UDP port number must be specified. A colon is used to separate the address and port, `IP:port` for IPv4 or
+`[IP]:port` for IPv6.
+
+#### Transport
+UDP is used to transport DNS messages and `Upstream` health checks.
+
+#### Health Checking Details
+
+Nameserver health checking is performed in-band. A health check is performed only when CoreDNS detects an error.
+The check runs in a loop, every 0.5s, for as long as the upstream reports unhealthy. Once healthy, CoreDNS stops
+health checking until the next error. The health checks use a recursive DNS query (. IN NS) to get upstream health.
+Any response that is not a network error (REFUSED, NOTIMPL, SERVFAIL, etc) is taken as a healthy upstream.
+When all `Nameservers` are down CoreDNS assumes health checking as a mechanism has failed and will try to connect to a
+random nameserver (which may or may not work).
+
+### Risks and Mitigations
```

Comment:
We need to check how this affects memory usage: AIUI, each upstream will require an additional instance of the `forward` plugin.  CoreDNS upstream has several open issues related to high memory usage when using many zones with certain plugins.  It's particularly an issue with such plugins as the `kubernetes` plugin, and if we're lucky, the `forward` plugin is lightweight enough that it will not pose a problem, but that is an assumption we need to validate.

Reply:
Sounds like something to call out in the test plan and/or Graduation Criteria? Throughout the development of the feature could we collect CI metrics for analysis to compare master with the other releases?

(To be clear, you're suggesting call-outs to inform the development going forward, _not_ suggesting we do this analysis before accepting the proposal?)

Reply:
Right, this is something we will need to look at when we implement the proposal, and based on our observations, we may need to increase resource limits (although we're now setting a limit of 512Mi, which I hope would be high enough: https://github.com/openshift/cluster-dns-operator/blob/master/assets/dns/daemonset.yaml#L49-L54) or warning in documentation if adding a large number of upstreams would incur a performance penalty or hit memory limits.  Now that you mention metrics, though, I'm wondering, do we publish a metric to track memory usage?